### PR TITLE
Compatibility with jit.script and torch.compile: COMPLETE

### PR DESCRIPTION
**Pull request by borisfom**

### CODE REVIEW

**Comment by borisfom** [comment]
Note: depends on backend MR #74.

File: cuequivariance_torch/cuequivariance_torch/primitives/tensor_product.py [hunk 1]
```diff
@@ -190,7 +240,7 @@ def _tensor_product_fx(
             seg_shape = descriptor.get_segment_shape(-1, path)
             outputs += [
                 out.reshape(
-                    out.shape[: out.ndim - len(seg_shape)] + (math.prod(seg_shape),)
+                    out.shape[: out.ndim - len(seg_shape)] + (prod(seg_shape),)
```
Comment (mariogeiger) [inline cuequivariance_torch/cuequivariance_torch/primitives/tensor_product.py:243]: I thought you said that you gave up on enabling script for the fx code. What is this for then?

File: cuequivariance_torch/cuequivariance_torch/primitives/symmetric_tensor_product.py [hunk 2]
```diff
@@ -335,9 +335,9 @@ def forward(
         i0 = i0.to(torch.int32)
         x0 = x0.reshape(x0.shape[0], x0.shape[1] // self.u, self.u)
         x1 = x1.reshape(x1.shape[0], x1.shape[1] // self.u, self.u)
-        logger.debug(
-            f"Calling SymmetricTensorContraction: {self.descriptors}, input shapes: {x0.shape}, {i0.shape}, {x1.shape}"
-        )
+        # logger.debug(
+        #     f"Calling SymmetricTensorContraction: {self.descriptors}, input shapes: {x0.shape}, {i0.shape}, {x1.shape}"
+        # )
```
Comment (mariogeiger) [inline cuequivariance_torch/cuequivariance_torch/primitives/symmetric_tensor_product.py:340]: Shall we do the if is scripting trick here too?

File: cuequivariance_torch/cuequivariance_torch/primitives/tensor_product.py (hunk 1 above)
Comment (borisfom) [reply cuequivariance_torch/cuequivariance_torch/primitives/tensor_product.py:243]: Oh it's in that part! Thanks, I will revert it.

File: cuequivariance_torch/cuequivariance_torch/primitives/symmetric_tensor_product.py (hunk 2 above)
Comment (borisfom) [reply cuequivariance_torch/cuequivariance_torch/primitives/symmetric_tensor_product.py:340]: Sure!

File: cuequivariance_torch/cuequivariance_torch/primitives/tensor_product.py (hunk 1 above)
Comment (borisfom) [reply cuequivariance_torch/cuequivariance_torch/primitives/tensor_product.py:243]: The ones in _reshape are actually used in FusedTensorProductOp3 - so probably best to keep them all consistent.

File: cuequivariance_torch/cuequivariance_torch/primitives/equivariant_tensor_product.py [hunk 3]
```diff
@@ -21,6 +21,58 @@
...
+    def forward(
+        self,
+        inputs: List[torch.Tensor],
+        indices: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        if indices is not None:
+            # TODO: at some point we will have kernel for this
+            assert len(inputs) >= 1
+            inputs[0] = inputs[0][indices]
+        return  self.tp(inputs)
```
[the remark this answers is not in the document]
Comment (borisfom) [reply cuequivariance_torch/cuequivariance_torch/primitives/equivariant_tensor_product.py:40]: So ?

File: cuequivariance_torch/cuequivariance_torch/primitives/equivariant_tensor_product.py [hunk 4]
```diff
@@ -21,6 +21,58 @@
...
+            if x0.shape[0] == 1:
+                indices = torch.zeros(
+                    (x1.shape[0],), dtype=torch.int32, device=x1.device
+                )
+            else:  #  x0.shape[0] == x1.shape[0]:
+                indices = torch.arange(
+                    x1.shape[0], dtype=torch.int32, device=x1.device
+                )
+        # borisf : why was it here ?
+        # if indices is not None:
```
[the remark this answers is not in the document]
Comment (borisfom) [reply cuequivariance_torch/cuequivariance_torch/primitives/equivariant_tensor_product.py:73]: Well I guess we should move that condition into the assert. Also, why branching at all? arange(1, ...) would give the same result as zeros((1,),...).

File: cuequivariance_torch/tests/primitives/equivariant_tensor_product_test.py [hunk 5]
```diff
@@ -12,15 +12,209 @@
...
 
+torch._dynamo.config.cache_size_limit = 100
+
+
+try:
+    import cuequivariance_ops_torch.onnx  # noqa: F401
+    import onnx  # noqa: F401
+    import onnxruntime  # noqa: F401
+    import onnxscript  # noqa: F401
+    from cuequivariance_ops_torch.tensorrt import register_plugins
```
Comment (mariogeiger) [inline cuequivariance_torch/tests/primitives/equivariant_tensor_product_test.py:34]: why `tensorrt` here in the onnx import block?

File: cuequivariance_torch/tests/operations/symmetric_contraction_test.py [hunk 6]
```diff
@@ -30,7 +32,7 @@
 @pytest.mark.parametrize("dtype", [torch.float64, torch.float32])
 @pytest.mark.parametrize("layout", [cue.ir_mul, cue.mul_ir])
 @pytest.mark.parametrize("original_mace", [True, False])
-@pytest.mark.parametrize("batch", [0, 32])
+@pytest.mark.parametrize("batch", [1, 32])
```
Comment (mariogeiger) [inline cuequivariance_torch/tests/operations/symmetric_contraction_test.py:35]: We must not forget to add back this test batch=0 when the backend properly supports it

**Review by mariogeiger — APPROVED** [review]
I think we can merge

**Pull request merged**